File: Kotlin/test/appendix/AtomicityLockingTests.kt
package appendix

import appendix.AtomicityLocking.Rank1
import appendix.AtomicityLocking.Rank2
import java.util.concurrent.Callable
import java.util.concurrent.CountDownLatch
import java.util.concurrent.Executors
import kotlin.math.max
import kotlin.test.Test
import kotlin.test.assertEquals

class AtomicityLockingTests {
  @Test
  fun testGetRank1() {
    testGetRank({ Rank1() }, { it.getRank() })
  }

  @Test
  fun testGetRank2() {
    testGetRank({ Rank2() }, { it.getRank() })
  }

  private fun <A> testGetRank(make: () -> A, getRank: (A) -> Int?) {
    val n = max(Runtime.getRuntime().availableProcessors(), 6)
    val exec = Executors.newCachedThreadPool()
    repeat(10_000) {
      val start = CountDownLatch(n)
      val rank = make()
      val tasks = List(n) {
        Callable {
          start.countDown()
          start.await()
          getRank(rank)
        }
      }
      val values = exec.invokeAll(tasks).map { it.get() }.toSet()
      assertEquals(setOf(1, 2, 3, 4, 5, null), values)
    }
    exec.shutdown()
  }
}

File: Scala/test/appendix/AtomicityLockingTests.scala
package appendix

import org.scalatest.funsuite.AnyFunSuite
import tinyscalautils.control.times
import tinyscalautils.threads.Executors.global
import tinyscalautils.threads.{ availableProcessors, countDownAndWait, withThreadPoolAndWait }

import java.util.concurrent.CountDownLatch
import scala.concurrent.Future
import scala.jdk.OptionConverters.RichOptionalInt

class AtomicityLockingTests extends AnyFunSuite:
   import AtomicityLockingJ.*

   private def testGetRank[A](make: () => A, getRank: A => Option[Int]) =
      val n        = availableProcessors.max(6)
      val expected = Set.tabulate(5)(x => Some(x + 1)) + None
      10_000 times {
         val set = withThreadPoolAndWait(global) {
            val start = CountDownLatch(n)
            val rank  = make()
            Future.traverse(Set.range(0, n)) { _ =>
               Future {
                  start.countDownAndWait()
                  getRank(rank)
               }
            }
         }
         assert(set == expected)
      }

   test("LockObject.getRank") {
      testGetRank(() => LockObject(), _.getRank().toScala)
   }

   test("LockMethod.getRank") {
      testGetRank(() => LockMethod(), _.getRank().toScala)
   }

   test("LockThis.getRank") {
      testGetRank(() => LockThis(), _.getRank().toScala)
   }
